### PR TITLE
Add ability to run topsapp steps for bursts

### PR DESCRIPTION
**Pull request by forrestfwilliams**
This will need to wait to be merged until the DEM, template, and ancillary data download features are added.

TODO:
- [x] Add ability to run steps
- [x] Add tests for step running
- [x] Add vrt swap functionality
- [x] Test vrt swap
- [x] Explore ability to turn off pickling (search "pickle" [here](https://github.com/isce-framework/isce2#running-isce))
- [x] Merge all upstream changes

### CODE REVIEW

File: tests/test_topsapp.py [hunk 1]
```diff
@@ -15,13 +19,39 @@ def test_topsapp_burst_config(tmp_path):
...
-        assert '[-118.0, 37.0, -117.0, 38.0]' in template
-        assert '[1]' in template
+    template = template_file.read()
+    assert 'S1A_IW_SLC__1SDV_20200604T022251_20200604T022318_032861_03CE65_7C85.SAFE' in template
+    assert 'S1A_IW_SLC__1SDV_20200616T022252_20200616T022319_033036_03D3A3_5D11.SAFE' in template
+    assert 'orbits' in template
+    assert 'aux_cal' in template
+    assert 'dem.tif' in template
+    assert '[-118.0, 37.0, -117.0, 38.0]' in template
+    assert '[1]' in template
```
Comment (jtherrmann) [inline tests/test_topsapp.py:31]:
```suggestion
            template = template_file.read()
            assert 'S1A_IW_SLC__1SDV_20200604T022251_20200604T022318_032861_03CE65_7C85.SAFE' in template
            assert 'S1A_IW_SLC__1SDV_20200616T022252_20200616T022319_033036_03D3A3_5D11.SAFE' in template
            assert 'orbits' in template
            assert 'aux_cal' in template
            assert 'dem.tif' in template
            assert '[-118.0, 37.0, -117.0, 38.0]' in template
            assert '[1]' in template
```

File: src/hyp3_isce2/topsapp.py [hunk 2]
```diff
@@ -81,3 +107,68 @@ def write_template(self, filename: Union[str, Path] = 'topsapp.xml') -> Path:
...
+    reference and secondary bursts with custom VRTs that point to the actual burst rasters.
+    """
+    ref_vrt_list = [str(x) for x in Path('reference').glob('**/*.vrt')]
+    sec_vrt_list = [str(x) for x in Path('secondary').glob('**/*.vrt')]
+    if len(ref_vrt_list) + len(sec_vrt_list) != 2:
+        raise ValueError(
+            (
+                'There should only be 2 VRT files in the reference and secondary directories',
+                ', check that you are performing a single burst run.',
+            )
```
Comment (jtherrmann) [inline src/hyp3_isce2/topsapp.py:125]:
```suggestion
            'There should only be 2 VRT files in the reference and secondary directories, '
            'check that you are performing a single burst run.',
```

File: src/hyp3_isce2/topsapp.py [hunk 3]
```diff
@@ -81,3 +107,66 @@ def write_template(self, filename: Union[str, Path] = 'topsapp.xml') -> Path:
...
+
+    To convince topsApp to process a burst pair, we need to swap the VRTs it generates for the
+    reference and secondary bursts with custom VRTs that point to the actual burst rasters.
+    """
+    ref_vrt_list = [str(x) for x in Path('reference').glob('**/*.vrt')]
+    sec_vrt_list = [str(x) for x in Path('secondary').glob('**/*.vrt')]
+    if len(ref_vrt_list) + len(sec_vrt_list) != 2:
+        raise ValueError(
+            'There should only be 2 VRT files in the reference and secondary directories, '
+            'check that you are performing a single burst run.',
```
Comment (jtherrmann) [inline src/hyp3_isce2/topsapp.py:123]:
```suggestion
            'check that you are performing a single burst run.'
```

**Comment by AndrewPlayer3** [comment]
For "Explore ability to turn off pickling (search "pickle" [here](https://github.com/isce-framework/isce2#running-isce))," turning off pickling should just mean running it without steps, correct? When I remove the steps option 
"+ step_args" at https://github.com/ASFHyP3/hyp3-isce2/blob/caffa525c5f1fb7f8eb44e41afcc088d87749ff5/src/hyp3_isce2/topsapp.py#L169, the tests all still pass. What sort of exploration should be done?

File: environment.yml [hunk 4]
```diff
@@ -3,10 +3,13 @@ channels:
...
   - nodefaults
 dependencies:
-  - python>=3.8
+  # python and numpy pins: https://github.com/ASFHyP3/hyp3-isce2/issues/14
+  - python>=3.8,<=3.10
+  - numpy<1.20
   - pip
-  - dem_stitcher>=2.4.0
   - isce2
+  - numpy
```
Comment (jtherrmann) [inline environment.yml:11]:
```suggestion
```

File: src/hyp3_isce2/process.py [hunk 5]
```diff
@@ -3,20 +3,25 @@
...
 from hyp3_isce2.burst import BurstParams, download_bursts, get_region_of_interest
+from hyp3_isce2.dem import download_dem_for_isce2
 from hyp3_isce2.s1_auxcal import download_aux_cal
-from hyp3_isce2.topsapp import TopsappBurstConfig
 
 
 log = logging.getLogger(__name__)
-TOPSAPP = str(Path(site.getsitepackages()[0]) / 'isce' / 'applications')
+
+ISCE_APPLICATIONS = Path(site.getsitepackages()[0]) / 'isce' / 'applications'
```
Comment (jtherrmann) [inline src/hyp3_isce2/process.py:22]: Can we leave a comment explaining why we have to do this?

File: src/hyp3_isce2/s1_auxcal.py [hunk 6]
```diff
@@ -32,6 +32,7 @@ def _download_platform(url: str, aux_cal_dir: Path):
         url: URL to download the aux cal files from.
         aux_cal_dir: Directory to download the aux cal files to.
     """
+    # TODO increases retries for this request: https://github.com/ASFHyP3/hyp3-isce2/issues/18
```
Comment (jtherrmann) [inline src/hyp3_isce2/s1_auxcal.py:35]:
```suggestion
    # TODO increase retries for this request: https://github.com/ASFHyP3/hyp3-isce2/issues/18
```

File: tests/test_topsapp.py [hunk 7]
```diff
@@ -23,5 +27,47 @@ def test_topsapp_burst_config(tmp_path):
...
+        swath=1,
+        azimuth_looks=1,
+        range_looks=1,
+    )
+    template_path = config.write_template(tmp_path / 'topsApp.xml')
+
+    with pytest.raises(ValueError):
+        run_topsapp_burst('notastep', config_xml=template_path)
+
+    with pytest.raises(ValueError):
```
Comment (jtherrmann) [inline tests/test_topsapp.py:69]:
```suggestion
    with pytest.raises(ValueError, match=r'^If dostep is specified, start and stop cannot be used$'):
```

File: src/hyp3_isce2/process.py [hunk 8]
```diff
@@ -3,20 +3,27 @@
...
 from hyp3_isce2.burst import BurstParams, download_bursts, get_region_of_interest
+from hyp3_isce2.dem import download_dem_for_isce2
 from hyp3_isce2.s1_auxcal import download_aux_cal
-from hyp3_isce2.topsapp import TopsappBurstConfig
 
 
 log = logging.getLogger(__name__)
-TOPSAPP = str(Path(site.getsitepackages()[0]) / 'isce' / 'applications')
+
+# ISCE needs it's applications to be on the system path.
```
Comment (jtherrmann) [inline src/hyp3_isce2/process.py:22]:
```suggestion
# ISCE needs its applications to be on the system path.
```

**Review by jtherrmann — APPROVED** [review]

**Pull request merged**